Mathlib/Analysis/Convex/Schur.lean

```diff
@@ -16,19 +16,65 @@ The proofs are adapted from https://www.cip.ifi.lmu.de/~grinberg/VornicuS.pdf
 
 lemma vornicu_schur_mildorf {α : Type*} [PartialOrder α] [CommRing α] [IsOrderedRing α]
     {a b c x y z : α}
-    (hx : 0 ≤ x) (hz : 0 ≤ z)
-    (hcb : c ≤ b) (hba : b ≤ a)
-    (hyxz : y ≤ x + z) :
+    (hx : 0 ≤ x) (hz : 0 ≤ z) (hcb : c ≤ b) (hba : b ≤ a) (hyxz : y ≤ x + z) :
     0 ≤ x * (a - b) * (a - c) + y * (b - a) * (b - c) + z * (c - a) * (c - b) := by
   have h₁ : 0 ≤ a - b := by linear_combination hba
   have h₂ : 0 ≤ b - c := by linear_combination hcb
-  have h₃ : 0 ≤ x + z - y := by linear_combination hyxz
-  have h₄ : 0 ≤ x * (a - b) := mul_nonneg hx h₁
-  calc
-    _ ≤ (x + z - y) * (b - c) * (a - b) := mul_nonneg (mul_nonneg h₃ h₂) h₁
-    _ = x * (a - b) * (b - c) + y * (b - a) * (b - c) + z * (a - b) * (b - c) := by ring
-    _ ≤ x * (a - b) * (a - c) + y * (b - a) * (b - c) + z * (a - c) * (b - c) := by gcongr
-    _ = x * (a - b) * (a - c) + y * (b - a) * (b - c) + z * (c - a) * (c - b) := by ring
+  linear_combination h₁ * x * (a - b) + h₂ * z * (b - c) + hyxz * (b - c) * (a - b)
+
+lemma schur {a b c r : ℝ} (ha : 0 ≤ a) (hb : 0 ≤ b) (hc : 0 ≤ c) (hr : 0 ≤ r) :
+    0 ≤ a ^ r * (a - b) * (a - c) + b ^ r * (b - a) * (b - c) + c ^ r * (c - a) * (c - b) := by
+  wlog hcb : c ≤ b generalizing a b c
+  next => linear_combination this ha hc hb (le_of_not_le hcb)
+  wlog hba : b ≤ a generalizing a b c
+  next =>
+    obtain hac | hca := le_total a c
+    case inl => linear_combination this hb hc ha hac hcb
+    case inr => linear_combination this hb ha hc hca (le_of_not_le hba)
+  refine vornicu_schur_mildorf (Real.rpow_nonneg ha _) (Real.rpow_nonneg hc _) hcb hba ?_
+  linear_combination Real.rpow_le_rpow hb hba hr + Real.rpow_nonneg hc r
+
+lemma sq_le_sq_of_nonpos {α : Type*} [LinearOrder α] [CommRing α] [IsOrderedRing α]
+    {a b : α} (hab : a ≤ b) (hb : b ≤ 0) : b ^ 2 ≤ a ^ 2 := by
+  rw [sq, sq]
+  exact mul_le_mul_of_nonpos_of_nonpos hab hab (hab.trans hb) hb
+
+lemma pow_le_pow_left_of_nonpos₀ {α : Type*} [LinearOrder α] [CommRing α] [IsOrderedRing α]
+    {a b : α} {n : ℕ} (hab : a ≤ b) (hb : b ≤ 0)
+    (hn : Even n) : b ^ n ≤ a ^ n := by
+  obtain ⟨n, rfl⟩ := hn
+  rw [← two_mul, pow_mul, pow_mul]
+  apply pow_le_pow_left₀ (sq_nonneg _)
+  exact sq_le_sq_of_nonpos hab hb
+
+lemma schur_pos {a b c r : ℝ} (ha : 0 < a) (hb : 0 < b) (hc : 0 < c) :
+    0 ≤ a ^ r * (a - b) * (a - c) + b ^ r * (b - a) * (b - c) + c ^ r * (c - a) * (c - b) := by
```

Comment:
Do you think the semiring version is remotely interesting?
```lean
    a ^ r * (a * c + b * a) + b ^ r * (b * c + a * b) + c ^ r * (c * b + a * c)
      ≤ a ^ r * (a * a + b * c) + b ^ r * (b * b + a * c) + c ^ r * (c * c + a * b) by
```

Reply:
Nope! I don't mind adding it anyway, although I'm not 100% sure it's true. (But maybe you can embed the semiring into a ring or something...)

Reply:
Actually this particular one uses rpow, so there's definitely not a semiring version here. I presume your actual intention is for the ones which currently use CommRing

Reply:
Well, you could write this one for `NNReal`!